shared/coordinator/src/data_selection.rs

```diff
@@ -10,20 +11,22 @@ pub fn assign_data_for_state<T: NodeIdentity>(
 ) -> BTreeMap<BatchId, T> {
     let round = coordinator.current_round().unwrap();
 
-    let trainer_nodes: Vec<_> = (0..coordinator.epoch_state.clients.len())
+    // Get trainer nodes references to avoid cloning
+    let mut trainer_nodes: Vec<_> = (0..coordinator.epoch_state.clients.len())
         .filter_map(|i| {
-            let client = &coordinator.epoch_state.clients[i];
+            let client = coordinator.epoch_state.clients.get(i)?;
```

Comment:
small nit, I think you can iterate directly over the clients and indicies with enumerate

```rust
coordinator.epoch_state.clients.iter().enumerate().filter_map(|(i, client)| ...)
```